File: examples/bluesolar_angles_python/BluesolarHelpers.py
import sys

# Ensure Python can find the compiled module
sys.path.append('C:/Users/manue_6t240gh/Dropbox/OpenSource/heliostat_tracking/build/Release')

import heliostat_tracking_module as ht
from math import radians

def convert_and_validate_arguments(args):
    if len(args) != 9:
        print("Usage: python script.py xHeliostat yHeliostat zHeliostat xAimingPoint yAimingPoint zAimingPoint sunAzimuth sunElevation")
        return False

    try:
        xHeliostat = float(args[1])
        yHeliostat = float(args[2])
        zHeliostat = float(args[3])
        xAimingPoint = float(args[4])
        yAimingPoint = float(args[5])
        zAimingPoint = float(args[6])
        sunAzimuth = float(args[7])
        sunElevation = float(args[8])
    except ValueError as e:
        print(f"Invalid argument: {e}")
        return False

    return xHeliostat, yHeliostat, zHeliostat, xAimingPoint, yAimingPoint, zAimingPoint, sunAzimuth, sunElevation

def bluesolar_heliostat_armature(armature_id):
    ht.set_primary_shift(armature_id, 0.0, 0.0, 2.415)
    ht.set_primary_axis(armature_id, -1.0, 0.0, 0.0)
    ht.set_primary_angles(armature_id, 0.0, 90.0)

    ht.set_secondary_shift(armature_id, 0.0, -0.0816, 0.0)
    ht.set_secondary_axis(armature_id, 0.0, 0.0, -1.0)
    ht.set_secondary_angles(armature_id, -70.0, 55.0)

    ht.set_facet_shift(armature_id, 0.0, -0.105, 0.035)
    ht.set_facet_normal(armature_id, 0.0, -1.0, 0.0)

    ht.set_angles_default(armature_id, 0.0, 0.0)

def create_bluesolar_elevation_angle_km():
    gamma = 1.5566945190927768  # Convert to radians
    rab = 0.3679941188052566
    rbc = 0.08587871211683747
    rad = 0.45469491923653677
    alpha2 = 0.08155204289894769
    offset = 0.04037
    return ht.ElevationAngleKM(gamma, rab, rbc, rad, alpha2, offset)

def create_bluesolar_hour_angle_km():
    gamma = 0.543717625543648  # Convert to radians
    rab = 0.3716059721933388
    rbc = 0.05209218963038278
    rad = 0.3390154085801952
    offset = 0.04037    
    return ht.HourAngleKM(gamma, rab, rbc, rad, offset)

File: examples/bluesolar_angles/BluesolarHelpers.cpp
#include <iomanip>
#include <iostream>

#include "TrackerArmature2A.h"
#include "TrackerTarget.h"
#include "Ray.h"
#include "Transform.h"
#include "ElevationAngleKM.h"
#include "HourAngleKM.h"
#include "BluesolarHelpers.h"

bool ConvertAndValidateArguments(int argc, char* argv[], 
                                 double& xHeliostat, double& yHeliostat, double& zHeliostat, 
                                 double& xAimingPoint, double& yAimingPoint, double& zAimingPoint, 
                                 double& sunAzimuth, double& sunElevation) {
    if (argc != 9) {
        std::cerr << "Usage: " << argv[0] << " xHeliostat yHeliostat zHeliostat xAimingPoint yAimingPoint zAimingPoint sunAzimuth sunElevation" << std::endl;
        return false;
    }

    char* end;
    xHeliostat = std::strtod(argv[1], &end);
    if (end == argv[1]) { std::cerr << "Invalid argument: " << argv[1] << std::endl; return false; }
    yHeliostat = std::strtod(argv[2], &end);
    if (end == argv[2]) { std::cerr << "Invalid argument: " << argv[2] << std::endl; return false; }
    zHeliostat = std::strtod(argv[3], &end);
    if (end == argv[3]) { std::cerr << "Invalid argument: " << argv[3] << std::endl; return false; }
    xAimingPoint = std::strtod(argv[4], &end);
    if (end == argv[4]) { std::cerr << "Invalid argument: " << argv[4] << std::endl; return false; }
    yAimingPoint = std::strtod(argv[5], &end);
    if (end == argv[5]) { std::cerr << "Invalid argument: " << argv[5] << std::endl; return false; }
    zAimingPoint = std::strtod(argv[6], &end);
    if (end == argv[6]) { std::cerr << "Invalid argument: " << argv[6] << std::endl; return false; }
    sunAzimuth = std::strtod(argv[7], &end);
    if (end == argv[7]) { std::cerr << "Invalid argument: " << argv[7] << std::endl; return false; }
    sunElevation = std::strtod(argv[8], &end);
    if (end == argv[8]) { std::cerr << "Invalid argument: " << argv[8] << std::endl; return false; }

    return true;
}

void BluesolarHeliostatArmature(TrackerArmature2A& armature)
{
    armature.set_primaryShift(vec3d(0.0, 0.0, 2.415)); // primary shift vector.
    armature.set_primaryAxis(vec3d(-1.0, 0.0, 0.0)) ; // unit vector in the direction of the first axis of rotation while in default position.
    armature.set_primaryAngles(vec2d(0.0, 90.0)); // min and max rotation angles around the primary axis of rotation in degrees.

    armature.set_secondaryShift(vec3d(0.0, -0.0816, 0.0)); // secondary shift vector.
    armature.set_secondaryAxis(vec3d(0.0, 0.0, -1.0)); // unit vector in the direction of the second axis of rotation.
    armature.set_secondaryAngles(vec2d(-70.0, 55.0)); // min and max rotation angles around the secondary axis of rotation in degrees.

    armature.set_facetShift(vec3d(0., -0.105, 0.035)); // facet shift vector.
    armature.set_facetNormal(vec3d(0.0, -1.0, 0.0)); // unit vector in the direction of the facet's normal when in default position.

    armature.set_anglesDefault(vec2d(0.0, 0.0)); // values of the rotation angles in the default position in degrees.
}

ElevationAngleKM CreateBluesolarElevationAngleKM() {
    double gamma = 1.499539835163685; // radians
    double rab = 0.38228347073744173; // meters
    double rbc = 0.0396; // meters
    double rad = 0.4146341554709371; // meters
    double alpha2 = 0.08480554835440447; // meters
    return ElevationAngleKM(gamma, rab, rbc, rad, alpha2);
}

HourAngleKM CreateBluesolarHourAngleKM() {
    double gamma = 0.6388776401148127; // radians
    double rab = 0.34805695317203444; // meters
    double rbc = 0.04225; // meters
    double rad = 0.33827680301912516; // meters
    return HourAngleKM(gamma, rab, rbc, rad);
}